src/NHibernate.Test/Async/TypesTest/TimestampTypeFixture.cs

```diff
@@ -21,4 +24,42 @@ public class TimestampTypeFixtureAsync : AbstractDateTimeTypeFixtureAsync
 		protected override string TypeName => "Timestamp";
 		protected override AbstractDateTimeType Type => NHibernateUtil.Timestamp;
 	}
+
+	// Testing SQL Server 2008 with datetime in db instead of datetime2
+	[TestFixture]
+	public class TimestampTypeMixedDateTimeFixtureAsync : TimestampTypeFixtureAsync
+	{
+		protected override bool AppliesTo(Dialect.Dialect dialect)
+		{
+			return base.AppliesTo(dialect) && dialect is MsSql2008Dialect;
+		}
+
+		protected override void Configure(Configuration configuration)
+		{
+			base.Configure(configuration);
+
+			configuration.SetProperty(Environment.SqlTypesKeepDateTime, "true");
+		}
+
+		/* Uncomment for testing how it fails if NHibernate uses DateTime2 while the db column is DateTime.
+		 * It fails two thirds of times on update with an undue optimistic locking exception, due to the datetimes
+		 * ending with 3.333ms (transmitted as 3) or 6.666ms (transmitted as 7), which does not match the datetime2
+		 * sent by NHibernate for checking the revision has not changed. Demonstrated by UpdateUseExpectedSqlType.
```

Comment:
If using a db with `datetime` but letting NHibernate use `datetime2`, versioning is broken. `SqlTypesKeepDateTime` must be used for supporting datetime based versioning. (Or better of course, migrate the db to `datetime2`.)